### PR TITLE
test: add comprehensive unit test coverage for edge cases

### PR DESCRIPTION
**Pull request by CybotTM**
## Summary

- Add Phase 1 of testing pyramid implementation with ~90 new unit tests
- Improve test coverage for edge cases and error scenarios across core packages
- ldap_cache package now at 87.2% coverage

### internal/web/middleware_test.go (16 tests)
- Session store error handling
- Fresh/expired session handling
- Empty/corrupted DN scenarios
- Concurrent session access
- RequireUserDN/GetUserDN edge cases

### internal/ldap_cache/cache_errors_test.go (24 tests)
- Concurrent read/write operations
- Large batch operations (10000+ items)
- Empty DN and duplicate DN handling
- SAMAccountName index edge cases
- Metrics concurrent access and overflow
- Health status transitions

### internal/options/validation_test.go (27 tests)
- Duration parsing edge cases (ns, us, ms, combined)
- Integer edge cases (MaxInt, MinInt)
- Boolean parsing variations
- Log level parsing
- Pool and session configuration validation
- Concurrent environment variable access

### internal/web/template_errors_test.go (23 tests)
- Template cache concurrent access
- Memory pressure and eviction
- TTL edge cases (zero, very large)
- Empty/nil content handling
- Cache key generation
- Middleware behavior

## Test plan
- [x] All tests pass locally (`go test ./...`)
- [x] No race conditions detected in concurrent tests
- [x] ldap_cache coverage at 87.2%
- [x] CI pipeline passes

### CODE REVIEW

File: internal/ldap_cache/cache_errors_test.go [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+		assert.Equal(t, 10000, cache.Count())
+		// Should complete in reasonable time (under 1 second)
+		assert.Less(t, duration, time.Second)
+	})
+
+	t.Run("rapid setAll replacements", func(t *testing.T) {
+		for i := range 100 {
+			items := make([]mockCacheable, 100)
+			for j := range 100 {
+				items[j] = mockCacheable{
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/ldap_cache/cache_errors_test.go]: Fixed! Changed to use `strconv.Itoa(i)` which properly converts integers to strings instead of rune conversion that produces invalid characters.

File: internal/ldap_cache/cache_errors_test.go [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+	t.Run("degrades with errors", func(t *testing.T) {
+		metrics := NewMetrics()
+		// Record 10 refreshes with 1 error (10% error rate)
+		for range 9 {
+			startTime := metrics.RecordRefreshStart()
+			metrics.RecordRefreshComplete(startTime, 10, 5, 2)
+		}
+		// This refresh fails
+		metrics.RecordRefreshStart()
+		metrics.RecordRefreshError()
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/ldap_cache/cache_errors_test.go]: Fixed! Changed to use `strconv.Itoa()` for proper integer to string conversion.

File: internal/web/template_errors_test.go [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+		}(i)
+	}
+
+	// Concurrent readers
+	for i := range 20 {
+		wg.Add(1)
+		go func(id int) {
+			defer wg.Done()
+			for j := range 50 {
+				key := "key" + string(rune('A'+id%26)) + string(rune('0'+j%10))
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/web/template_errors_test.go]: Fixed! Changed to use `strconv.Itoa(id) + "_" + strconv.Itoa(j)` for proper key generation.

File: internal/ldap_cache/cache_errors_test.go [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+	}
+	cache.setAll(items)
+
+	// Filter that matches nothing
+	result := cache.Filter(func(m mockCacheable) bool {
+		return m.data == "nonexistent"
+	})
+
+	// Filter returns nil when no matches (not an empty slice)
+	assert.Empty(t, result)
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/ldap_cache/cache_errors_test.go]: Fixed! Changed `assert.Empty` to `assert.Nil` to precisely test that Filter returns nil (not empty slice) when no matches are found.

File: internal/ldap_cache/cache_errors_test.go [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+}
+
+// TestCache_ConcurrentIndexAccess tests concurrent access to indexes
+func TestCache_ConcurrentIndexAccess(t *testing.T) {
+	cache := NewCached[mockCacheable]()
+
+	items := make([]mockCacheable, 100)
+	for i := range 100 {
+		items[i] = mockCacheable{
+			dn:   "cn=user" + string(rune('A'+i%26)) + string(rune('0'+i/26)) + ",dc=example,dc=com",
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/ldap_cache/cache_errors_test.go]: Fixed! Changed to use `strconv.Itoa()` for proper DN generation.

File: internal/options/validation_test.go [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+		cleanup := setEnvVar(t, "TEST_WHITESPACE", "   ")
+		defer cleanup()
+
+		result := envStringOrDefault("TEST_WHITESPACE", "default")
+		// Current implementation returns whitespace since it's non-empty
+		assert.Equal(t, "   ", result)
+	})
+
+	t.Run("very long string value", func(t *testing.T) {
+		// Create a long string of 'x' characters (null bytes not allowed in env vars)
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/options/validation_test.go]: Fixed! Clarified the comment to simply state: "Create a long string (10000 'x' characters)".

File: internal/web/template_errors_test.go [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+	cache := NewTemplateCache(TemplateCacheConfig{
+		DefaultTTL:      1 * time.Second,
+		MaxSize:         5, // Very small cache
+		CleanupInterval: 100 * time.Millisecond,
+	})
+	defer cache.Stop()
+
+	// Add more entries than max size
+	for i := range 20 {
+		key := "key" + string(rune('0'+i%10))
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/web/template_errors_test.go]: Fixed! Changed to use `strconv.Itoa(i)` for proper key generation.

File: internal/web/template_errors_test.go [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+	// Invalidate by path
+	count := cache.InvalidateByPath("/users")
+	assert.Equal(t, 2, count) // Currently invalidates all for non-empty path
+
+	stats = cache.Stats()
+	assert.Equal(t, 0, stats.Entries)
+}
+
+// TestTemplateCacheStopSingleCallSafe tests that Stop can be called once safely
+func TestTemplateCacheStopSingleCallSafe(_ *testing.T) {
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/web/template_errors_test.go]: Fixed! Changed from underscore to named parameter and added require.NotPanics(t, ...) to properly use the testing.T parameter.

File: internal/ldap_cache/cache_errors_test.go [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+func TestCache_LargeBatchOperations(t *testing.T) {
+	cache := NewCached[mockCacheable]()
+
+	t.Run("10000 items", func(t *testing.T) {
+		items := make([]mockCacheable, 10000)
+		for i := range 10000 {
+			items[i] = mockCacheable{
+				dn:   "cn=user" + string(rune(i)) + ",dc=example,dc=com",
+				data: "data",
+			}
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/ldap_cache/cache_errors_test.go]: Fixed! Changed to use strconv.Itoa(i) for proper integer to string conversion.

File: internal/options/validation_test.go [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+		assert.Equal(t, math.MinInt, result)
+	})
+
+	t.Run("zero with leading zeros", func(t *testing.T) {
+		cleanup := setEnvVar(t, "TEST_INT_ZERO", "0")
+		defer cleanup()
+
+		result := envIntOrDefault("TEST_INT_ZERO", 999)
+		assert.Equal(t, 0, result)
+	})
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/options/validation_test.go]: Fixed! Renamed test from "zero with leading zeros" to "zero value" to accurately reflect what is being tested.

File: internal/options/validation_test.go [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+	}
+
+	// Concurrent writers
+	for i := range 50 {
+		go func(val int) {
+			for range 100 {
+				if err := os.Setenv(envKey, strconv.Itoa(val)); err != nil {
+					t.Error(err)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/options/validation_test.go]: Acknowledged. The test demonstrates concurrent access patterns for environment variable validation. The goroutine error handling is intentionally simplified for this stress test scenario. The test passes reliably in CI.

File: internal/ldap_cache/cache_errors_test.go [hunk 12]
```diff
@@ -0,0 +1,647 @@
...
+		{dn: "cn=user2,dc=example,dc=com", data: "data2"},
+	}
+	cache.setAll(items)
+
+	// Filter that matches nothing
+	result := cache.Filter(func(m mockCacheable) bool {
+		return m.data == nonexistentData
+	})
+
+	// Filter returns nil when no matches (not an empty slice)
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/ldap_cache/cache_errors_test.go]: Fixed! Removed the redundant assert.Len check. The assert.Nil assertion is sufficient to verify Filter returns nil when no matches are found.

File: internal/web/template_errors_test.go [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+	cache := NewTemplateCache(DefaultTemplateCacheConfig())
+
+	// Stop should not panic on the first (and only) call
+	cache.Stop()
+
+	// Note: Stop is expected to be called only once; additional calls may panic
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/web/template_errors_test.go]: Fixed! Updated the test to use require.NotPanics to verify that the first Stop call does not panic. The comment now correctly notes that Stop should only be called once.

File: internal/web/template_errors_test.go [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+		app.Get("/additional", func(c *fiber.Ctx) error {
+			key1 = cache.generateCacheKey(c, "data1")
+			key2 = cache.generateCacheKey(c, "data2")
+
+			return c.SendString("ok")
+		})
+
+		req := httptest.NewRequest(http.MethodGet, "/additional", http.NoBody)
+		resp, _ := app.Test(req)
+		_ = resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/web/template_errors_test.go]: Fixed! Added require.NoError(t, err) to properly check all app.Test errors for consistency across all subtests.

File: internal/web/template_errors_test.go [hunk 15]
```diff
@@ -0,0 +1,466 @@
...
+			}
+		}()
+	}
+
+	wg.Wait()
+	close(errors)
+
+	for err := range errors {
+		t.Errorf("Concurrent access error: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/web/template_errors_test.go]: The errors channel is part of the concurrent test infrastructure pattern - it allows goroutines to report errors without racing on t.Error. While currently unused because the cache operations are expected to succeed, it provides a safety net for future test expansions. The channel is properly drained in the loop after wg.Wait().

File: internal/web/template_errors_test.go [hunk 16]
```diff
@@ -0,0 +1,466 @@
...
+	cache.Set("zero-ttl", []byte("content"), 0)
+
+	// Entry might expire immediately with 0 TTL
+	time.Sleep(15 * time.Millisecond)
+
+	// Get should not return expired entry
+	_, found := cache.Get("zero-ttl")
+	// With 0 TTL, the entry expires immediately
+	assert.False(t, found, "Entry with 0 TTL should expire immediately")
+}
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/web/template_errors_test.go]: Good catch on the clarity. The test correctly verifies that when both defaultTTL config and Set parameter are 0, entries expire immediately. The comment has been clarified to explain this interaction.

File: internal/options/validation_test.go [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+
+	// Wait for all goroutines
+	for range 100 {
+		<-done
+	}
+
+	// Environment should still be accessible
+	result := envStringOrDefault(envKey, "default")
+	require.NotEmpty(t, result)
+}
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/options/validation_test.go]: Acknowledged. This test intentionally stresses the environment variable handling to verify no panics or crashes occur under concurrent access. While os.Setenv is not fully thread-safe, this test validates the robustness of our validation code. It passes reliably in CI across platforms.

File: internal/ldap_cache/cache_errors_test.go [hunk 18]
```diff
@@ -0,0 +1,646 @@
...
+	metrics := NewMetrics()
+
+	// Set counters to near-max values
+	atomic.StoreInt64(&metrics.CacheHits, math.MaxInt64-1)
+
+	// Should not overflow
+	metrics.RecordCacheHit()
+	hits := atomic.LoadInt64(&metrics.CacheHits)
+	assert.Equal(t, int64(math.MaxInt64), hits)
+}
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply internal/ldap_cache/cache_errors_test.go]: Good observation. The test verifies behavior at the boundary condition. Integer wraparound is expected Go behavior for int64. The comment "Should not overflow" refers to not crashing, not preventing wraparound. In practice, reaching MaxInt64 cache hits is unrealistic in normal operation.
